### PR TITLE
[runtime] Fixed allowing non-object complex arguments to cgt.un

### PR DESCRIPTION
**Pull request by alexanderkyte**

### CODE REVIEW

**Comment by monoadmin** [comment]
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.

File: mono/metadata/verify.c [hunk 1]
```diff
@@ -3058,7 +3058,10 @@ do_cmp_op (VerifyContext *ctx, const unsigned char table [TYPE_MAX][TYPE_MAX], g
 	a = stack_pop (ctx);
 
 	if (opcode == CEE_CGT_UN) {
-		if (stack_slot_get_type (a) == TYPE_COMPLEX && stack_slot_get_type (b) == TYPE_COMPLEX) {
+		// page 302 of ECMA-335, subheading 2
```
Comment (tritao) [inline mono/metadata/verify.c]: IMHO it's best to actually quote the relevant pieces of the standard.

**Comment by alexanderkyte** [comment]
Is it preferred step here to amend and force push to my fork or to make a new commit?

**Comment by alexrp** [comment]
Amend and force push to keep the history clean.

**Comment by tritao** [comment]
whitelist

File: mono/metadata/verify.c [hunk 2]
```diff
@@ -3058,7 +3058,12 @@ do_cmp_op (VerifyContext *ctx, const unsigned char table [TYPE_MAX][TYPE_MAX], g
 	a = stack_pop (ctx);
 
 	if (opcode == CEE_CGT_UN) {
-		if (stack_slot_get_type (a) == TYPE_COMPLEX && stack_slot_get_type (b) == TYPE_COMPLEX) {
+		// ECMA-335:
+		// cgt.un is allowed and verifiable on ObjectRefs (O). This is commonly used when
+		// comparing an ObjectRef with null
+		if (stack_slot_get_type (a) == TYPE_COMPLEX && a->type->type == MONO_TYPE_OBJECT &&
+			stack_slot_get_type (b) == TYPE_COMPLEX && b->type->type == MONO_TYPE_OBJECT ) {
```
Comment (alexrp) [inline mono/metadata/verify.c]: nit: no space before closing parenthesis

Comment (alexanderkyte) [inline mono/metadata/verify.c]: Nit has been picked

**Comment by alexrp** [comment]
Had to revert this as it broke the build. See 29a4519 for details.

**Pull request merged**